packages/@aws-cdk/aws-ec2/lib/vpc-flow-logs.ts

```diff
@@ -0,0 +1,240 @@
+import iam = require('@aws-cdk/aws-iam');
+import logs = require('@aws-cdk/aws-logs');
+import s3 = require('@aws-cdk/aws-s3');
+import { Construct, IResource, Resource } from '@aws-cdk/core';
+import { CfnFlowLog } from './ec2.generated';
+import { ISubnet, IVpc } from './vpc';
+
+export interface IVpcFlowLog extends IResource {
+  /**
+   * The Id of the VPC Flow Log
+   */
+  readonly flowLogId: string;
+}
+
+/**
+ * The type of VPC traffic to log
+ */
+export enum VpcFlowLogTrafficType {
+  /**
+   * Only log accepts
+   */
+  ACCEPT = 'ACCEPT',
+
+  /**
+   * Log all requests
+   */
+  ALL = 'ALL',
+
+  /**
+   * Only log rejects
+   */
+  REJECT = 'REJECT'
+}
+
+/**
+ * The type of resource to create the flow log for
+ */
+enum VpcFlowLogResourceType {
+  NETWORK_INTERFACE = 'NetworkInterface',
+  SUBNET = 'Subnet',
+  VPC = 'VPC'
+}
+
+/**
+ * The destination type for the flow log
+ */
+export enum VpcFlowLogDestinationType {
+  /**
+   * CloudWatch logs destination
+   */
+  CLOUD_WATCH_LOGS = 'cloud-watch-logs',
+
+  /**
+   * S3 destination
+   */
+  S3 = 's3'
+}
+
+/**
+ * Options for the type of resource you
+ * are creating the flow log for
+ */
+export interface ResourceTypeOptions {
+  /**
+   * The network interface Id that you want to
+   * create the flow log for
+   */
+  readonly networkInterface?: string;
+
+  /**
+   * The VPC Subnet that you want to create the
+   * flow log for
+   */
+  readonly subnet?: ISubnet;
+
+  /**
+   * The VPC that you want to create the flow log for
+   */
+  readonly vpc?: IVpc;
+}
+
+/**
+ * Options to add a flow log to a VPC
+ */
+export interface VpcFlowLogOptions {
+  /**
+   * The options for creating a CloudWatch log group
+   *
+   * @default - will use logs.LogGroupProps defaults
+   */
+  readonly logGroupOptions?: logs.LogGroupProps;
+
+  /**
+   * The type of traffic to log. You can log traffic that the resource accepts or rejects, or all traffic.
+   *
+   * @default ALL
+   */
+  readonly trafficType?: VpcFlowLogTrafficType;
+
+  /**
+   * S3 bucket to publish the flow logs to
+   *
+   * @default - if destinationType is S3 then and this is not
+   * provided then it will create a bucket for you. 
+   *
+   * Should not be provided if destinationType is CLOUD_WATCH_LOGS
+   */
+  readonly s3Bucket?: s3.IBucket;
+
+  /**
+   * Specifies the type of destination to which the flow log data is to be published. 
+   * Flow log data can be published to CloudWatch Logs or Amazon S3
+   *
+   * @default VpcFlowLogDestinationType.CLOUD_WATCH_LOGS
+   */
+  readonly destinationType?: VpcFlowLogDestinationType;
```

Comment:
I think destination can also use a union-like pattern. It will dramatically simplify your validation and the code within the FlowLog constructor.

Reply:
I tried to follow some other examples of where this is done, but I'm not sure if what I did is the ideal way of doing things.